File: scripts/py/validate_negbin.py
#!/usr/bin/env python3
"""
Validate Tutorial 4: Negative Binomial GLM coefficients

This script verifies the coefficients shown in the Negative Binomial tutorial
match actual Python output using the UCI Bike Sharing dataset.

Expected coefficients (from tutorial):
  Intercept: 8.1969
  temp: 1.7005
  hum: -0.4825
  windspeed: -1.1239
  workingday: 0.0519
  weathersit: -0.1421
  theta: 6.773
"""

import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf

# Load the UCI Bike Sharing dataset (daily aggregates)
bike = pd.read_csv("../../data/day.csv")

print(f"Dataset dimensions: {bike.shape}")
print(f"\nCount summary:\n{bike['cnt'].describe()}")

# For Negative Binomial in statsmodels, alpha = 1/theta
# With theta = 6.773, alpha = 1/6.773 = 0.1476
expected_theta = 6.773
alpha = 1 / expected_theta

# Fit Negative Binomial model
model = smf.glm(
    'cnt ~ temp + hum + windspeed + workingday + weathersit',
    data=bike,
    family=sm.families.NegativeBinomial(alpha=alpha)
).fit()

# Display results
print("\n=== Negative Binomial GLM Validation ===\n")
print("Model Summary:")
print(model.summary())

print("\n\nCoefficients:")
print(model.params)

print(f"\n\nAlpha (1/theta) used: {alpha:.4f}")
print(f"Equivalent theta: {1/alpha:.3f}")

# Compare to expected values
expected = {
    "Intercept": 8.1969,
    "temp": 1.7005,
    "hum": -0.4825,
    "windspeed": -1.1239,
    "workingday": 0.0519,
    "weathersit": -0.1421
}

actual = model.params.round(4)

print("\n\nValidation Check:")
print("Expected vs Actual (Coefficients):")
for name, exp_val in expected.items():
    act_val = actual.get(name, float('nan'))
    match = abs(exp_val - act_val) < 0.01
    status = "OK" if match else "MISMATCH"
    print(f"  {name}: expected {exp_val:.4f}, got {act_val:.4f} [{status}]")

# Compare with Poisson for SE differences
print("\n\nComparison with Poisson (SE ratios):")
fit_pois = smf.glm(
    'cnt ~ temp + hum + windspeed + workingday + weathersit',
    data=bike,
    family=sm.families.Poisson()
).fit()

for name in expected.keys():
    se_pois = fit_pois.bse.get(name, float('nan'))
    se_nb = model.bse.get(name, float('nan'))
    ratio = se_nb / se_pois
    print(f"  {name}: Poisson SE={se_pois:.4f}, NegBin SE={se_nb:.4f}, Ratio={ratio:.1f}x")

# Model fit statistics
print("\n\nModel Fit Statistics:")
print(f"  Deviance: {model.deviance:.2f}")
print(f"  Deviance/df: {model.deviance / model.df_resid:.2f} (should be ~1 for good fit)")
print(f"  AIC (NegBin): {model.aic:.2f}")
print(f"  AIC (Poisson): {fit_pois.aic:.2f}")

if __name__ == "__main__":
    pass  # Script runs on import

File: scripts/R/validate-negbin.R
# Validate Tutorial 4: Negative Binomial GLM coefficients
# This script verifies the coefficients shown in the Negative Binomial tutorial
# match actual R output using the UCI Bike Sharing dataset.
#
# Expected coefficients (from tutorial):
#   Intercept: 8.1969
#   temp: 1.7005
#   hum: -0.4825
#   windspeed: -1.1239
#   workingday: 0.0519
#   weathersit: -0.1421
#   theta: 6.773

library(MASS)

# Load the UCI Bike Sharing dataset (daily aggregates)
bike <- read.csv("../../data/day.csv")

cat("Dataset dimensions:", dim(bike), "\n")
cat("Count summary:\n")
print(summary(bike$cnt))

# Fit Negative Binomial model using glm.nb from MASS
model <- glm.nb(
  cnt ~ temp + hum + windspeed + workingday + weathersit,
  data = bike
)

# Display results
cat("\n=== Negative Binomial GLM Validation ===\n\n")
cat("Model Summary:\n")
print(summary(model))

cat("\n\nCoefficients:\n")
print(coef(model))

cat("\n\nTheta (dispersion parameter):", model$theta, "\n")
cat("SE of Theta:", model$SE.theta, "\n")

# Compare to expected values
expected <- c(
  "(Intercept)" = 8.1969,
  "temp" = 1.7005,
  "hum" = -0.4825,
  "windspeed" = -1.1239,
  "workingday" = 0.0519,
  "weathersit" = -0.1421
)

actual <- round(coef(model), 4)

cat("\n\nValidation Check:\n")
cat("Expected vs Actual (Coefficients):\n")
for (name in names(expected)) {
  act_val <- actual[name]
  match <- abs(expected[name] - act_val) < 0.01
  status <- if (match) "OK" else "MISMATCH"
  cat(sprintf("  %s: expected %.4f, got %.4f [%s]\n",
              name, expected[name], act_val, status))
}

# Check theta
expected_theta <- 6.773
match_theta <- abs(expected_theta - model$theta) < 0.1
cat(sprintf("  theta: expected %.3f, got %.3f [%s]\n",
            expected_theta, model$theta, if (match_theta) "OK" else "MISMATCH"))

# Compare with Poisson for SE differences
cat("\n\nComparison with Poisson (SE ratios):\n")
fit_pois <- glm(cnt ~ temp + hum + windspeed + workingday + weathersit,
                family = poisson(link = "log"), data = bike)

se_pois <- summary(fit_pois)$coefficients[, 2]
se_nb <- summary(model)$coefficients[, 2]
se_ratio <- se_nb / se_pois

for (name in names(se_ratio)) {
  cat(sprintf("  %s: Poisson SE=%.4f, NegBin SE=%.4f, Ratio=%.1fx\n",
              name, se_pois[name], se_nb[name], se_ratio[name]))
}

# Model fit statistics
cat("\n\nModel Fit Statistics:\n")
cat(sprintf("  Null deviance: %.2f\n", model$null.deviance))
cat(sprintf("  Residual deviance: %.2f\n", model$deviance))
cat(sprintf("  Deviance/df: %.2f (should be ~1 for good fit)\n",
            model$deviance / model$df.residual))
cat(sprintf("  AIC (NegBin): %.2f\n", AIC(model)))
cat(sprintf("  AIC (Poisson): %.2f\n", AIC(fit_pois)))
